personalized_nlp/learning/regressor.py

```diff
@@ -44,6 +42,26 @@ def training_step(self, batch, batch_idx):
 
     def training_epoch_end(self, outs):
         pass
+    
+    def freeze(self) -> None:
+        for name, param in self.named_parameters():
+            if 'fc' not in name: 
+                param.requires_grad = False
+        self._frozen = True
+
+    def unfreeze(self) -> None:
+        if self._frozen:
+            for name, param in self.named_parameters():
+                if 'fc' not in name: 
+                    param.requires_grad = True
+        self._frozen = False
+
+    def on_epoch_start(self):
+        if self.current_epoch < self.hparams.nr_frozen_epochs:
+            self.freeze()
+
+        if self.current_epoch >= self.hparams.nr_frozen_epochs:
+            self.unfreeze()
```

Comment:
In the final version we will move it to clabback, for now it stays here because it is ok, i.e., it works and we ought to focus on experiments.

Reply:
Yes, I have fixed the logic of the `on_epoch_start` hook a little bit, and verify that also. If we set `nr_frozen_epoch = 0`, and `frozen=False`, we will finetune from scratch, if  we set `nr_frozen_epoch = epochs`, and `frozen=True` we will freeze all as before. So for now I think we can run the experiment on it, can you double-check on it also?

Reply:
> nr_frozen_epoch = 0 and frozen = False -- we finetune from the first epoch, but to make it work i think that there should be if self.current_epoch < self.hparams.nr_frozen_epochs: in 60th line because in other if we start to count from 0 so at the beggining self.current_epoch is 0 and it will freeze layers even if we do not want it.

- Yes, it's correct. I've fixed that also XD after I commented here I've just recognized it.
For the entire I think it'll work like that.